### PR TITLE
PLANNER-2754 Add CS-B map() implementation

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

File: core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package org.optaplanner.constraint.streams.bavet.common;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import org.optaplanner.constraint.streams.bavet.uni.UniTuple;
+import org.optaplanner.constraint.streams.bavet.uni.UniTupleImpl;
+
+public abstract class AbstractMapNode<InTuple_ extends Tuple, Right_>
```
Comment (triceo) [inline core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java:9]: Here is the meat of the PR.

File: core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/bi/BavetMapBridgeBiConstraintStream.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public ConstraintStream getTupleSource() {
+        return parent.getTupleSource();
+    }
+
+    @Override
+    public <Score_ extends Score<Score_>> void buildNode(NodeBuildHelper<Score_> buildHelper) {
+        if (!childStreamList.isEmpty()) {
+            throw new IllegalStateException("Impossible state: the stream (" + this
+                    + ") has an non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```
Comment (rsynek) [inline core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/bi/BavetMapBridgeBiConstraintStream.java:56]:
```suggestion
                    + ") has a non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```

File: core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/quad/BavetMapBridgeQuadConstraintStream.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public ConstraintStream getTupleSource() {
+        return parent.getTupleSource();
+    }
+
+    @Override
+    public <Score_ extends Score<Score_>> void buildNode(NodeBuildHelper<Score_> buildHelper) {
+        if (!childStreamList.isEmpty()) {
+            throw new IllegalStateException("Impossible state: the stream (" + this
+                    + ") has an non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```
Comment (rsynek) [inline core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/quad/BavetMapBridgeQuadConstraintStream.java:56]:
```suggestion
                    + ") has a non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```

File: core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/tri/BavetMapBridgeTriConstraintStream.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public ConstraintStream getTupleSource() {
+        return parent.getTupleSource();
+    }
+
+    @Override
+    public <Score_ extends Score<Score_>> void buildNode(NodeBuildHelper<Score_> buildHelper) {
+        if (!childStreamList.isEmpty()) {
+            throw new IllegalStateException("Impossible state: the stream (" + this
+                    + ") has an non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```
Comment (rsynek) [inline core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/tri/BavetMapBridgeTriConstraintStream.java:56]:
```suggestion
                    + ") has a non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```

File: core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/uni/BavetMapBridgeUniConstraintStream.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public ConstraintStream getTupleSource() {
+        return parent.getTupleSource();
+    }
+
+    @Override
+    public <Score_ extends Score<Score_>> void buildNode(NodeBuildHelper<Score_> buildHelper) {
+        if (!childStreamList.isEmpty()) {
+            throw new IllegalStateException("Impossible state: the stream (" + this
+                    + ") has an non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```
Comment (rsynek) [inline core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/uni/BavetMapBridgeUniConstraintStream.java:55]:
```suggestion
                    + ") has a non-empty childStreamList (" + childStreamList + ") but it's a flattenLast bridge.");
```

File: core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    protected abstract Right_ map(InTuple_ tuple);
+
+    @Override
+    public void update(InTuple_ tuple) {
+        UniTupleImpl<Right_> outTuple = tuple.getStore(inputStoreIndex);
+        if (outTuple == null) {
+            // No fail fast if null because we don't track which tuples made it through the filter predicate(s)
+            insert(tuple);
```
Comment (rsynek) [inline core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java:48]: Why do we insert the tuple in this case?

**Review by rsynek — APPROVED** [review]
LGTM. One question and a couple of small suggestions inline.

File: core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java (hunk 6 above)
Comment (triceo) [reply core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java:48]: This is a bit of a quirk of the filter implementation. Insert can be totally filtered out, they will not propagate. Which means that if we're seeing the tuple for the first time doing update, it actually means an insert.

Comment (rsynek) [reply core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java:48]: Ok. Maybe worth mentioning in the comment (or perhaps it's clear for those more familiar with Bavet)?

Comment (triceo) [reply core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractMapNode.java:48]: Being familiar with Bavet, the existing comment makes sense to me.
That said, there should eventually be a more general documentation for Bavet which would explain the inner workings - this being one of them.

**Pull request merged**